otx/algorithms/segmentation/adapters/mmseg/models/utils/proto_utils.py

```diff
@@ -0,0 +1,131 @@
+"""Utils for prototype learning."""
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+import math
+import warnings
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+
+def distributed_sinkhorn(out, sinkhorn_iterations=3, epsilon=0.05):
+    """Sinkhorn distribution."""
+    L = torch.exp(out / epsilon).t()  # K x B
+    B = L.shape[1]
+    K = L.shape[0]
+
+    # make the matrix sums to 1
+    sum_L = torch.sum(L)
+    L /= sum_L
+
+    for _ in range(sinkhorn_iterations):
+        L /= torch.sum(L, dim=1, keepdim=True)
+        L /= K
+
+        L /= torch.sum(L, dim=0, keepdim=True)
+        L /= B
+
+    L *= B
+    L = L.t()
+
+    indexs = torch.argmax(L, dim=1)
+    L = F.gumbel_softmax(L, tau=0.5, hard=True)
+
+    return L, indexs
+
+
+def l2_normalize(x):
+    """L2 normalization."""
+    return F.normalize(x, p=2, dim=-1)
+
+
+def momentum_update(old_value, new_value, momentum, debug=False):
+    """EMA update function."""
+    update = momentum * old_value + (1 - momentum) * new_value
+    if debug:
+        print(
+            "old prot: {:.3f} x |{:.3f}|, new val: {:.3f} x |{:.3f}|, result= |{:.3f}|".format(
+                momentum,
+                torch.norm(old_value, p=2),
+                (1 - momentum),
+                torch.norm(new_value, p=2),
+                torch.norm(update, p=2),
+            )
+        )
+    return update
+
+
+class ProjectionHead(nn.Module):
+    """Projection head to transfrom features space used further for prototype learning."""
+
+    def __init__(self, dim_in, proj_dim=256):
+        super(ProjectionHead, self).__init__()
+
+        self.proj = nn.Sequential(nn.Conv2d(dim_in, dim_in, 1), nn.ReLU(inplace=True), nn.Conv2d(dim_in, proj_dim, 1))
+
+    def forward(self, x):
+        """Farward method."""
+        return l2_normalize(self.proj(x))
```

Comment:
I think it's okay to refactor it F.normalize(x, p=2, dim=-1) to reduce call depth.

Reply:
Agree